// @progress/kendo-react-common/tree-utils/itemIdUtils.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
const c = "", b = "0", r = "_";
function S(t, n, o) {
  if (x(t))
    return n[Number(t)];
  {
    const e = n[Number(f(t))], s = e && e[o] || [];
    return s.length ? S(i(t), s, o) : void 0;
  }
}
function f(t) {
  return a(t) ? t : t.split(r)[0];
}
function i(t) {
  if (a(t))
    return t;
  {
    const n = t.indexOf(r);
    return t.substring(n + 1);
  }
}
function E(t) {
  return O("0", t);
}
function O(t, n) {
  return t = t.toString(), n ? n + r + t : t;
}
function h(t) {
  const n = t.lastIndexOf(r);
  return n < 0 ? c : t.substring(0, n);
}
function a(t) {
  return t === c || t.indexOf(r) < 0;
}
function x(t) {
  return t !== c && t.indexOf(r) < 0;
}
function P(t) {
  return t.split(r);
}
function g(t) {
  const n = t.lastIndexOf(r);
  return n < 0 ? t : t.substring(n + 1);
}
function p(t) {
  return g(t) === b;
}
function N(t, n) {
  const o = n;
  t = "r" + r + t, n = "r" + r + n;
  const e = h(t) + r;
  if (n.startsWith(e)) {
    const s = n.substring(e.length);
    if (s) {
      const u = f(s);
      if (Number(g(t)) < Number(u)) {
        const l = e + (Number(u) - 1).toString() + s.substring(u.length);
        return i(l);
      }
    }
  }
  return o;
}
export {
  c as EMPTY_ID,
  r as SEPARATOR,
  b as ZERO_LEVEL_ZERO_NODE_ID,
  O as createId,
  P as getAllShortIds,
  N as getDecrementedItemIdAfterRemoval,
  h as getDirectParentId,
  E as getFirstChildId,
  i as getIdWithoutRootParentId,
  S as getItemById,
  f as getRootParentId,
  g as getShortId,
  a as isIdEmptyOrZeroLevel,
  x as isIdZeroLevel,
  p as isItemFirstFromSiblings
};
