// @progress/kendo-react-layout/breadcrumb/BreadcrumbLink.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import a from "prop-types";
import { validatePackage as o, dispatchEvent as l, useDir as k, getTabIndex as f, classNames as d } from "@progress/kendo-react-common";
import { packageMetadata as C } from "../package-metadata.mjs";
const r = t.forwardRef((e, s) => {
  o(C);
  const c = t.useRef(null), n = t.useRef(null), m = t.useCallback(
    () => {
      n.current && n.current.focus();
    },
    [n]
  );
  t.useImperativeHandle(c, () => ({
    element: n.current,
    focus: m,
    props: e
  })), t.useImperativeHandle(s, () => c.current);
  const u = t.useCallback(
    (i) => {
      e.id && l(e.onItemSelect, i, i.target, { id: e.id });
    },
    [e.onItemSelect]
  ), b = t.useCallback(
    (i) => {
      e.id && l(e.onKeyDown, i, i.target, { id: e.id });
    },
    [e.onKeyDown]
  );
  return /* @__PURE__ */ t.createElement(
    "span",
    {
      "aria-current": e.ariaCurrent ? e.ariaCurrent : e.isLast,
      role: "link",
      id: e.id,
      ref: n,
      style: e.style,
      dir: k(n, e.dir),
      tabIndex: f(e.tabIndex, e.disabled),
      className: d(
        {
          "k-breadcrumb-root-link": e.isFirst,
          "k-breadcrumb-link": !e.isFirst,
          "k-breadcrumb-icontext-link": e.iconClass !== void 0 && e.text,
          "k-breadcrumb-icon-link": e.iconClass !== void 0 && !e.text,
          "k-disabled": e.disabled
        }
      ),
      onClick: u,
      onKeyDown: b
    },
    e.iconClass ? /* @__PURE__ */ t.createElement("span", { className: d(
      "k-icon",
      e.iconClass
    ) }) : e.icon ? e.icon : "",
    e.text && /* @__PURE__ */ t.createElement("span", { className: "k-breadcrumb-item-text" }, e.text)
  );
}), x = {
  id: a.string,
  className: a.string,
  tabIndex: a.number,
  style: a.object,
  dir: a.string,
  disabled: a.bool,
  text: a.string,
  icon: a.node,
  iconClass: a.string,
  onClick: a.func,
  ariaCurrent: a.bool
};
r.displayName = "KendoReactBreadcrumbLink";
r.propTypes = x;
export {
  r as BreadcrumbLink
};
