// @progress/kendo-react-layout/breadcrumb/Breadcrumb.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as r from "react";
import i from "prop-types";
import { packageMetadata as M } from "../package-metadata.mjs";
import { validatePackage as T, useDir as O, classNames as w, kendoThemeMaps as x, dispatchEvent as v } from "@progress/kendo-react-common";
import { BreadcrumbListItem as B } from "./BreadcrumbListItem.mjs";
import { BreadcrumbDelimiter as R } from "./BreadcrumbDelimiter.mjs";
import { BreadcrumbLink as K } from "./BreadcrumbLink.mjs";
import { BreadcrumbOrderedList as N } from "./BreadcrumbOrderedList.mjs";
const f = r.forwardRef((e, E) => {
  T(M);
  const c = r.useRef(null), n = r.useRef(null), g = r.useMemo(
    () => e.breadcrumbOrderedList || N,
    [e.breadcrumbOrderedList]
  ), F = r.useMemo(
    () => e.breadcrumbListItem || B,
    [e.breadcrumbListItem]
  ), S = r.useMemo(
    () => e.breadcrumbDelimiter || R,
    [e.breadcrumbDelimiter]
  ), L = r.useMemo(
    () => e.breadcrumbLink || K,
    [e.breadcrumbLink]
  ), D = r.useCallback(
    () => {
      n.current && n.current.focus();
    },
    [n]
  ), h = r.useMemo(
    () => e.disabled || !1,
    [e.disabled]
  );
  r.useImperativeHandle(c, () => ({
    element: n.current,
    focus: D,
    props: e
  })), r.useImperativeHandle(E, () => c.current);
  const u = O(n, e.dir), y = (t) => {
    c.current && v(e.onItemSelect, t, t.target, { id: t.target.id });
  }, I = (t) => {
    c.current && v(e.onKeyDown, t, t.target, { id: t.target.id });
  }, k = e.valueField || s.valueField, C = e.iconField || s.iconField, m = e.iconClassField || s.iconClassField, o = e.textField || s.textField, b = e.size || "medium";
  return /* @__PURE__ */ r.createElement(
    "nav",
    {
      "aria-label": e.ariaLabel,
      id: e.id,
      style: e.style,
      ref: n,
      dir: u,
      className: w(
        "k-breadcrumb k-breadcrumb-wrap",
        {
          "k-rtl": u === "rtl",
          "k-disabled": h,
          "k-breadcrumb-md": !e.size,
          [`k-breadcrumb-${x.sizeMap[b] || b}`]: b
        },
        e.className
      )
    },
    /* @__PURE__ */ r.createElement(g, { rootItem: !0 }, /* @__PURE__ */ r.createElement(r.Fragment, null, e.data.map((t, a, l) => {
      const d = t[k];
      if (a === 0)
        return /* @__PURE__ */ r.createElement(
          F,
          {
            key: d,
            isFirstItem: !0,
            isLastItem: l.length - 1 === a
          },
          /* @__PURE__ */ r.createElement(
            L,
            {
              isLast: l.length - 1 === a,
              isFirst: !0,
              id: String(d),
              icon: t[C] || void 0,
              iconClass: t[m] ? String(t[m]) : void 0,
              text: t[o] ? String(t[o]) : void 0,
              disabled: t.disabled || !1,
              onItemSelect: y,
              onKeyDown: I,
              ...e
            }
          )
        );
    }))),
    /* @__PURE__ */ r.createElement(g, { rootItem: !1 }, /* @__PURE__ */ r.createElement(r.Fragment, null, e.data.map((t, a, l) => {
      const d = t[k];
      if (a !== 0)
        return /* @__PURE__ */ r.createElement(
          F,
          {
            key: d,
            isFirstItem: !1,
            isLastItem: l.length - 1 === a
          },
          /* @__PURE__ */ r.createElement(S, { dir: u }),
          /* @__PURE__ */ r.createElement(
            L,
            {
              isLast: l.length - 1 === a,
              isFirst: !1,
              id: String(d),
              icon: t[C] || void 0,
              iconClass: t[m] ? String(t[m]) : void 0,
              text: t[o] ? String(t[o]) : void 0,
              disabled: t.disabled || !1,
              onItemSelect: y,
              onKeyDown: I,
              ...e
            }
          )
        );
    })))
  );
}), z = {
  id: i.string,
  style: i.object,
  className: i.string,
  breadcrumbOrderedList: i.elementType,
  breadcrumbListItem: i.elementType,
  breadcrumbDelimiter: i.elementType,
  breadcrumbLink: i.elementType,
  data: i.arrayOf(i.shape({
    id: i.string,
    text: i.string,
    icon: i.any,
    iconClass: i.string
  })),
  dir: i.oneOf(["ltr", "rtl"]),
  disabled: i.bool,
  valueField: i.string,
  textField: i.string,
  iconField: i.string,
  iconClassField: i.string,
  onItemSelect: i.func,
  ariaLabel: i.string
}, s = {
  valueField: "id",
  textField: "text",
  iconField: "icon",
  iconClassField: "iconClass"
};
f.displayName = "KendoReactBreadcrumb";
f.propTypes = z;
f.defaultProps = s;
export {
  f as Breadcrumb
};
