// @progress/kendo-react-animation/Expand.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as r from "react";
import t from "prop-types";
import { Animation as c } from "./Animation.mjs";
const e = class e extends r.Component {
  /**
   * @hidden
   */
  render() {
    const {
      direction: n,
      children: o,
      ...a
    } = this.props, s = {
      transitionName: `expand-${n}`
    };
    return /* @__PURE__ */ r.createElement(c, { ...s, ...a }, o);
  }
};
e.propTypes = {
  children: t.oneOfType([
    t.arrayOf(t.node),
    t.node
  ]),
  childFactory: t.any,
  className: t.string,
  direction: t.oneOf([
    "horizontal",
    "vertical"
  ]),
  component: t.string,
  id: t.string,
  style: t.any
}, e.defaultProps = {
  appear: !1,
  enter: !0,
  exit: !0,
  transitionEnterDuration: 300,
  transitionExitDuration: 300,
  direction: "vertical"
};
let i = e;
export {
  i as Expand
};
