// @progress/kendo-react-animation/Zoom.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as i from "react";
import t from "prop-types";
import { Animation as a } from "./Animation.mjs";
const p = { position: "absolute", top: "0", left: "0" }, e = class e extends i.Component {
  /**
   * @hidden
   */
  render() {
    const {
      children: r,
      direction: o,
      ...s
    } = this.props;
    return /* @__PURE__ */ i.createElement(
      a,
      {
        ...s,
        transitionName: `zoom-${o}`,
        animationExitingStyle: this.props.stackChildren ? p : void 0
      },
      r
    );
  }
};
e.propTypes = {
  children: t.oneOfType([
    t.arrayOf(t.node),
    t.node
  ]),
  childFactory: t.any,
  className: t.string,
  direction: t.oneOf([
    "in",
    "out"
  ]),
  component: t.string,
  id: t.string,
  style: t.any,
  stackChildren: t.bool
}, e.defaultProps = {
  appear: !1,
  enter: !0,
  exit: !0,
  transitionEnterDuration: 300,
  transitionExitDuration: 300,
  direction: "out",
  stackChildren: !1
};
let n = e;
export {
  n as Zoom
};
