// @progress/kendo-react-layout/card/Card.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import r from "prop-types";
import { validatePackage as R, classNames as T } from "@progress/kendo-react-common";
import { cardType as c, cardOrientation as s } from "./interfaces/Enums.mjs";
import { packageMetadata as g } from "../package-metadata.mjs";
const a = e.forwardRef((t, l) => {
  R(g);
  const { children: d, dir: m, style: p, className: f, type: o, orientation: u, ...y } = t, n = e.useRef(null), i = e.useRef(null);
  return e.useImperativeHandle(n, () => ({
    element: i.current,
    props: t
  })), e.useImperativeHandle(l, () => n.current), /* @__PURE__ */ e.createElement(
    "div",
    {
      dir: m,
      style: p,
      ref: i,
      className: T(
        "k-card",
        f,
        { [`k-card-${o}`]: o !== c.DEFAULT },
        u !== s.HORIZONTAL ? "k-card-vertical" : "k-card-horizontal"
      ),
      ...y
    },
    d
  );
});
a.displayName = "KendoReactCard";
a.propTypes = {
  children: r.node,
  className: r.string,
  dir: r.string,
  orientation: r.oneOf(["horizontal", "vertical"]),
  style: r.object,
  type: r.oneOf(["default", "primary", "info", "success", "warning", "error"])
};
a.defaultProps = {
  type: c.DEFAULT,
  orientation: s.VERTICAL
};
export {
  a as Card
};
