// @progress/kendo-react-layout/contextmenu/ContextMenu.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as n from "react";
import { getActiveElement as u, classNames as s } from "@progress/kendo-react-common";
import { Popup as l } from "@progress/kendo-react-popup";
import { Menu as a } from "../menu/components/Menu.mjs";
const r = (e) => {
  const o = n.useRef(null);
  return n.useEffect(() => {
    e.show && setTimeout(() => {
      const m = o.current && o.current.element, t = m && m.querySelector(".k-menu-item"), c = u(t && t.ownerDocument);
      t && t !== c && t.focus();
    });
  }, [e.show]), /* @__PURE__ */ n.createElement(
    l,
    {
      show: e.show,
      offset: e.offset,
      popupClass: "k-menu-popup",
      animate: !1
    },
    /* @__PURE__ */ n.createElement(
      a,
      {
        vertical: !0,
        ...e,
        ref: o,
        className: s(
          "k-context-menu",
          e.className
        ),
        onClose: e.onClose,
        role: "menu"
      },
      e.children
    )
  );
};
r.displayName = "KendoReactContextMenu";
export {
  r as ContextMenu
};
