// @progress/kendo-react-layout/tilelayout/ResizeHandlers.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as n from "react";
import { Draggable as l } from "@progress/kendo-react-common";
const r = (t) => /* @__PURE__ */ n.createElement(
  l,
  {
    onPress: t.onPress,
    onDrag: (e) => {
      t.onResize(e.event, !1, t.d);
    },
    onRelease: (e) => {
      t.onResize(e.event, !0, t.d);
    }
  },
  /* @__PURE__ */ n.createElement("div", { className: "k-resize-handle k-cursor-" + t.d + "-resize", style: { bottom: 0, right: 0, ...t.style } })
);
class h extends n.Component {
  constructor() {
    super(...arguments), this.handleResize = (e, s, i) => {
      e.originalEvent.preventDefault(), this.props.onResize(e, { end: s, direction: i });
    };
  }
  render() {
    const { resizable: e, onPress: s, rtl: i } = this.props;
    return e ? /* @__PURE__ */ n.createElement(n.Fragment, null, e !== "vertical" && /* @__PURE__ */ n.createElement(
      r,
      {
        onPress: s,
        onResize: this.handleResize,
        d: "ew",
        style: i ? { top: 0, width: 9, left: 0, right: "" } : { top: 0, width: 9, right: 0, left: "" }
      }
    ), e !== "horizontal" && /* @__PURE__ */ n.createElement(r, { onPress: s, onResize: this.handleResize, d: "ns", style: { left: 0, height: 9 } }), e === !0 && /* @__PURE__ */ n.createElement(
      r,
      {
        onPress: s,
        onResize: this.handleResize,
        d: i ? "nesw" : "nwse",
        style: i ? { width: 9, height: 9, right: "", left: 0 } : { width: 9, height: 9, right: 0, left: "" }
      }
    )) : null;
  }
}
export {
  h as ResizeHandlers
};
