// @progress/kendo-react-common/hooks/useDroppable.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import { noop as n } from "../noop.mjs";
import { useInheritedState as k } from "./useInheritedState.mjs";
import { DropsContext as v, DropContext as E } from "../drag-n-drop/context/index.mjs";
import { useIsomorphicLayoutEffect as L } from "./use-isomorphic-layout-effect.mjs";
function y(e, m = {
  onDragEnter: n,
  onDragOver: n,
  onDragLeave: n,
  onDrop: n
}) {
  const {
    onDragEnter: s = n,
    onDragOver: c = n,
    onDragLeave: l = n,
    onDrop: u = n
  } = m, [, o] = k(E), [, D, g] = t.useContext(v), i = t.useCallback(
    () => e.current && e.current.element ? e.current.element : e.current,
    [e]
  ), a = t.useRef(null);
  t.useImperativeHandle(a, () => ({
    element: i(),
    onDragEnter: C,
    onDragOver: d,
    onDragLeave: b,
    onDrop: h,
    data: e.current
  }));
  const C = t.useCallback(
    (r) => {
      o(a, { target: e.current, event: r }), s(r);
    },
    [o, e, s]
  ), d = t.useCallback(
    (r) => {
      c(r);
    },
    [c]
  ), b = t.useCallback(
    (r) => {
      o(null, { target: e.current, event: r }), l(r);
    },
    [o, e, l]
  ), h = t.useCallback(
    (r) => {
      o(null, { target: e.current, event: r }), u(r);
    },
    [o, e, u]
  ), p = t.useCallback(
    () => (D(a), () => {
      g(a);
    }),
    [g, D]
  );
  L(p, [p]);
}
export {
  y as useDroppable
};
