// @progress/kendo-react-layout/timeline/TimelineHorizontal.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as n from "react";
import ft from "prop-types";
import { TimelineCard as ut } from "./TimelineCard.mjs";
import { addYearsFlags as mt } from "./utils.mjs";
import { caretAltLeftIcon as dt, caretAltRightIcon as kt } from "@progress/kendo-svg-icons";
import { useInternationalization as bt } from "@progress/kendo-react-intl";
import { Navigation as Et, canUseDOM as Tt, classNames as _, SvgIcon as z } from "@progress/kendo-react-common";
const vt = (v) => {
  const M = bt(), [y, j] = n.useState(!1), [R, K] = n.useState(!0), [X, Y] = n.useState(), [f, S] = n.useState(0), [g, q] = n.useState(), [l, B] = n.useState(0), [d, G] = n.useState(0), [a, O] = n.useState(1), [u, w] = n.useState(), [c, J] = n.useState(), [N, L] = n.useState([0, 0, 0]), [I, W] = n.useState(0), [Q, Z] = n.useState(1), m = n.useRef(null), i = n.useRef(null), D = n.useRef(null), { eventsData: tt, dateFormat: et, navigatable: nt, onActionClick: st } = v, U = g ? `${g}%` : "150px", k = n.useRef(new Et({
    tabIndex: 0,
    root: m,
    selectors: [".k-timeline-scrollable-wrap"]
  }));
  n.useEffect(() => {
    m.current && v.navigatable && (setTimeout(() => {
      const t = k.current.first;
      t && t.setAttribute("tabindex", String(0));
    }, 0), k.current.keyboardEvents = {
      keydown: {
        ArrowRight: at,
        ArrowLeft: ot,
        End: ct,
        Home: it
      }
    });
  }, [v.navigatable, a, l]), n.useEffect(() => {
    k.current.update();
  });
  const rt = (t) => {
    nt && k.current && k.current.triggerKeyboardEvent(t);
  };
  n.useEffect(() => {
    var r;
    const t = Tt && window.ResizeObserver && new window.ResizeObserver(p), e = i.current, s = ((r = m == null ? void 0 : m.current) == null ? void 0 : r.offsetWidth) || 0;
    return t && e && t.observe(e), x(), p(), L([s, 0, -s]), () => {
      t && t.disconnect();
    };
  }, []), n.useEffect(() => {
    const t = i.current, e = t && t.children[a];
    if (e) {
      const s = e.offsetWidth, r = !(f >= 0);
      if (j(r), c) {
        const o = c.length * s > l * s * ((f * -1 + 100) / 100);
        K(o);
      }
      (t == null ? void 0 : t.offsetWidth) * -f / 100 >= t.children.length * s && (x(), p(), S((e.offsetLeft - e.offsetWidth) / (l * e.offsetWidth) * 100 * -1)), V(e);
    }
  }, [l, g, a]);
  const at = (t, e, s) => {
    if (i.current) {
      const r = i.current.children[a + 1], o = r && r.classList.contains("k-timeline-flag-wrap") ? a + 2 : a + 1;
      if (c && c.length <= o)
        return;
      l + d <= o && P(), E(o, s);
    }
  }, ot = (t, e, s) => {
    if (i.current) {
      const r = i.current && i.current.children[a - 1], o = r && r.classList.contains("k-timeline-flag-wrap") ? a - 2 : a - 1;
      if (o <= 0)
        return;
      o < d && F(), E(o, s);
    }
  }, it = (t, e, s) => {
    if (i.current) {
      if (a <= 1)
        return;
      const r = 1;
      h(!0, r, 0, 0), E(r, s);
    }
  }, ct = (t, e, s) => {
    if (i.current) {
      const r = i.current && i.current.children.length - 1;
      if (r <= a)
        return;
      const o = Math.floor(r / l), T = o * -100, C = o * l;
      h(!0, r, C, T), E(r, s);
    }
  }, x = () => {
    const t = mt(tt);
    J(t), w([t[a - 1], t[a], t[a + 1]]);
  }, F = () => {
    var e;
    if (i.current && c) {
      const s = d - l, r = d - 1, o = c[r] && ((e = c[r]) != null && e.yearFlag) ? r - 1 : r, T = f + 100 > 0 ? 0 : f + 100;
      h(!1, o, s, T);
    }
  }, P = () => {
    if (i.current && c) {
      const e = d + l, s = c[e] && c[e].yearFlag ? e + 1 : e, r = f - 100;
      h(!0, s, e, r);
    }
  }, h = (t, e, s, r) => {
    A(t, e), p(), G(s), S(r), O(e);
  }, A = (t, e) => {
    if (u && c) {
      const s = b(t, Q), r = Object.assign([], u, { [s]: c[e] });
      w(r), Z(s), W(t ? -100 : 100), setTimeout(() => {
        var H;
        const o = (H = m.current) == null ? void 0 : H.offsetWidth, T = Object.assign([], N, {
          [s]: 0,
          [b(!0, s)]: o,
          [b(!1, s)]: o && -o
        }), C = Object.assign([], r, {
          [b(!0, s)]: {},
          [b(!1, s)]: {}
        });
        w(C), L(T), W(0);
      }, v.transitionDuration || 300);
    }
  }, b = (t, e) => {
    const s = t ? e + 1 : e - 1;
    return u ? s < 0 ? u.length - 1 : s % u.length : 0;
  }, V = (t) => {
    var s;
    const e = (a - d) * t.offsetWidth + t.offsetWidth / 2 + (((s = D.current) == null ? void 0 : s.offsetWidth) || 0);
    Y(e);
  }, E = (t, e) => {
    e.preventDefault(), t > a ? A(!0, t) : t < a && A(!1, t), O(t);
  }, p = n.useCallback(() => {
    const t = i.current, e = t && t.children[a];
    if (t instanceof HTMLElement && e) {
      const s = Math.round(t.offsetWidth / 150);
      V(e), q(100 / s), B(s);
    }
  }, []), lt = () => ({
    transitionDuration: "300ms"
  }), $ = (t) => ({
    transform: `translateX(${t}%)`,
    transformOrigin: "left top"
  });
  return /* @__PURE__ */ n.createElement(n.Fragment, null, /* @__PURE__ */ n.createElement("div", { className: "k-timeline-track-wrap", ref: m, onKeyDown: rt }, /* @__PURE__ */ n.createElement(
    "button",
    {
      ref: D,
      "aria-hidden": "true",
      onClick: () => y && F(),
      className: _(
        "k-timeline-arrow",
        "k-timeline-arrow-left",
        "k-button k-button-md k-rounded-md k-button-solid k-button-solid-base k-icon-button",
        { "k-disabled": !y }
      )
    },
    /* @__PURE__ */ n.createElement(z, { icon: dt })
  ), /* @__PURE__ */ n.createElement(
    "button",
    {
      "aria-hidden": "true",
      onClick: () => R && P(),
      className: _(
        "k-timeline-arrow",
        "k-timeline-arrow-right",
        "k-button k-button-md k-rounded-md k-button-solid k-button-solid-base k-icon-button",
        { "k-disabled": !R }
      )
    },
    /* @__PURE__ */ n.createElement(z, { icon: kt })
  ), /* @__PURE__ */ n.createElement("div", { className: "k-timeline-track" }, /* @__PURE__ */ n.createElement(
    "ul",
    {
      ref: i,
      className: "k-timeline-scrollable-wrap",
      role: "tablist",
      tabIndex: 0,
      style: { transform: `translateX(${f}%)` }
    },
    c && c.map((t, e) => t.yearFlag ? /* @__PURE__ */ n.createElement(
      "li",
      {
        role: "none",
        className: "k-timeline-track-item k-timeline-flag-wrap",
        style: { flex: `1 0 ${U}` },
        key: e
      },
      /* @__PURE__ */ n.createElement("span", { className: "k-timeline-flag" }, t.yearFlag)
    ) : /* @__PURE__ */ n.createElement(
      "li",
      {
        role: "tab",
        className: _(
          "k-timeline-track-item",
          { "k-focus": e === a }
        ),
        "aria-selected": e === a,
        style: { flex: `1 0 ${U}` },
        key: e
      },
      /* @__PURE__ */ n.createElement("div", { className: "k-timeline-date-wrap" }, /* @__PURE__ */ n.createElement("span", { className: "k-timeline-date" }, M.formatDate(t.date, et))),
      /* @__PURE__ */ n.createElement(
        "span",
        {
          onClick: (s) => E(e, s),
          className: "k-timeline-circle"
        }
      )
    ))
  ))), /* @__PURE__ */ n.createElement("div", { className: "k-timeline-events-list" }, /* @__PURE__ */ n.createElement(
    "ul",
    {
      className: "k-timeline-scrollable-wrap",
      style: I === 0 ? { ...$(I) } : { ...$(I), ...lt() }
    },
    u && u.map(
      (t, e) => /* @__PURE__ */ n.createElement(
        "li",
        {
          key: e,
          className: "k-timeline-event",
          style: { transform: `translate3d(${N[e]}px, 0px, 0px)` }
        },
        t && /* @__PURE__ */ n.createElement(
          ut,
          {
            tabindex: N[e] === 0 ? 0 : -1,
            eventData: t,
            onActionClick: st,
            horizontal: !0,
            collapsible: !1,
            calloutStyle: { left: `${X}px` }
          }
        )
      )
    )
  )));
};
vt.propTypes = {
  onActionClick: ft.func
};
export {
  vt as TimelineHorizontal
};
