// @progress/kendo-react-layout/timeline/Timeline.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import v from "prop-types";
import { classNames as h } from "@progress/kendo-react-common";
import { TimelineHorizontal as y } from "./TimelineHorizontal.mjs";
import { TimelineVertical as E } from "./TimelineVertical.mjs";
const T = (m) => {
  const p = {
    alterMode: !1,
    collapsibleEvents: !1,
    dateFormat: "MMM dd, yyyy",
    ...m
  }, {
    collapsibleEvents: a,
    transitionDuration: n,
    className: f,
    alterMode: r,
    navigatable: l,
    horizontal: t,
    events: o,
    dateFormat: s,
    onChange: d,
    onActionClick: c
  } = p, [k, u] = e.useState(), i = e.useRef(null);
  return e.useEffect(() => {
    i.current && u(i.current.offsetWidth);
  }, []), /* @__PURE__ */ e.createElement(
    "div",
    {
      ref: i,
      className: h(
        "k-timeline",
        {
          "k-timeline-collapsible": a,
          "k-timeline-vertical": !t,
          "k-timeline-alternating": r,
          "k-timeline-horizontal": t
        },
        f
      ),
      style: { width: `${k}px` }
    },
    t ? /* @__PURE__ */ e.createElement(
      y,
      {
        navigatable: l,
        eventsData: o,
        dateFormat: s,
        transitionDuration: n,
        onActionClick: c
      }
    ) : /* @__PURE__ */ e.createElement(
      E,
      {
        navigatable: l,
        eventsData: o,
        dateFormat: s,
        alterMode: r,
        collapsibleEvents: a,
        transitionDuration: n,
        onChange: d,
        onActionClick: c
      }
    )
  );
};
T.propTypes = {
  className: v.string
};
export {
  T as Timeline
};
