// @progress/kendo-react-progressbars/progressbar/ProgressBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import a from "prop-types";
import { useAnimation as H } from "@progress/kendo-react-animation";
import { validatePackage as K, useRtl as q, classNames as R, getTabIndex as G } from "@progress/kendo-react-common";
import { truncateNumber as J, updateProgress as f, calculatePercentage as v } from "../common/utils.mjs";
import { usePrevious as Q } from "./hooks/usePrevious.mjs";
import { DEFAULT_ANIMATION_DURATION as X, NO_ANIMATION as Y } from "../common/constants.mjs";
import { packageMetadata as Z } from "../package-metadata.mjs";
const k = e.forwardRef((t, S) => {
  K(Z);
  const {
    animation: b,
    disabled: g,
    reverse: C = p.reverse,
    orientation: A,
    labelVisible: T,
    labelPlacement: d,
    max: l = p.max,
    min: n = p.min,
    tabIndex: L,
    className: O,
    style: V,
    emptyStyle: h,
    emptyClassName: N,
    progressStyle: w,
    progressClassName: y
  } = t, o = e.useRef(null), i = e.useRef(null), c = e.useRef(null), U = e.useCallback(
    () => {
      o.current && o.current.focus();
    },
    []
  );
  e.useImperativeHandle(
    S,
    () => ({
      element: o.current,
      progressStatusElement: i.current,
      progressStatusWrapElement: c.current,
      focus: U
    })
  );
  const r = t.value || p.value, m = Q(r), P = t.value === null, D = q(o, t.dir), s = A === "vertical", M = J(r), _ = {
    value: r
  }, E = T ? t.label ? /* @__PURE__ */ e.createElement("span", { className: "k-progress-status" }, /* @__PURE__ */ e.createElement(t.label, { ..._ })) : /* @__PURE__ */ e.createElement("span", { className: "k-progress-status" }, M) : void 0, j = {
    className: R(
      "k-progressbar",
      {
        "k-progressbar-horizontal": !s,
        "k-progressbar-vertical": s,
        "k-progressbar-reverse": C,
        "k-progressbar-indeterminate": P,
        "k-disabled": g
      },
      O
    ),
    ref: o,
    dir: D,
    tabIndex: G(L, g),
    role: "progressbar",
    "aria-label": t.ariaLabel,
    "aria-valuemin": n,
    "aria-valuemax": l,
    "aria-valuenow": P ? void 0 : r,
    "aria-disabled": g,
    style: V
  }, x = R("k-progress-status-wrap", {
    "k-progress-start": d === "start",
    "k-progress-center": d === "center",
    "k-progress-end": d === "end" || d === void 0
  }), I = typeof b != "boolean" && b !== void 0 ? b.duration : b ? X : Y, B = e.useCallback(
    () => {
      const u = v(n, l, m);
      f(i, c, u, s);
    },
    [s, l, n, m]
  ), W = e.useCallback(
    (u) => {
      const F = v(n, l, m + (r - m) * u);
      f(i, c, F, s);
    },
    [n, l, m, r, s]
  ), z = e.useCallback(
    () => {
      const u = v(n, l, r);
      f(i, c, u, s);
    },
    [s, l, n, r]
  );
  return H(
    {
      duration: I,
      onStart: B,
      onUpdate: W,
      onEnd: z
    },
    [r, I]
  ), /* @__PURE__ */ e.createElement("div", { ...j }, /* @__PURE__ */ e.createElement("span", { className: x + (N ? " " + N : ""), style: h }, E), /* @__PURE__ */ e.createElement(
    "div",
    {
      className: "k-progressbar-value k-selected",
      style: w,
      ref: i
    },
    /* @__PURE__ */ e.createElement(
      "span",
      {
        className: x + (y ? " " + y : ""),
        ref: c
      },
      E
    )
  ));
});
k.propTypes = {
  animation: a.any,
  ariaLabel: a.string,
  disabled: a.bool,
  reverse: a.bool,
  label: a.any,
  labelVisible: a.bool,
  labelPlacement: a.oneOf(["start", "center", "end"]),
  max: a.number,
  min: a.number,
  value: a.number,
  tabIndex: a.number,
  emptyStyle: a.object,
  emptyClassName: a.string,
  progressStyle: a.object,
  progressClassName: a.string
};
const p = {
  animation: !1,
  min: 0,
  max: 100,
  value: 0,
  disabled: !1,
  reverse: !1,
  labelVisible: !0
};
k.defaultProps = p;
k.displayName = "KendoProgressBar";
export {
  k as ProgressBar
};
