// @progress/kendo-react-layout/appbar/AppBarSection.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import t from "prop-types";
import { focusFirstFocusableChild as u, classNames as f } from "@progress/kendo-react-common";
const r = e.forwardRef((n, o) => {
  const {
    children: l,
    className: a,
    style: i
  } = n, s = e.useRef(null), c = e.useCallback(
    () => {
      s.current && u(s.current);
    },
    []
  ), m = e.useCallback(
    () => ({
      element: s.current,
      focus: c
    }),
    [c]
  );
  e.useImperativeHandle(o, m);
  const p = e.useMemo(
    () => f(
      "k-appbar-section",
      a
    ),
    [a]
  );
  return /* @__PURE__ */ e.createElement("div", { className: p, style: i }, l);
});
r.propTypes = {
  children: t.any,
  className: t.string,
  style: t.object
};
r.displayName = "KendoAppBarSection";
export {
  r as AppBarSection
};
