// @progress/kendo-react-layout/bottomnavigation/BottomNavigationItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import t from "prop-types";
import { validatePackage as K, classNames as D, svgIconPropType as T, IconWrap as h } from "@progress/kendo-react-common";
import { packageMetadata as B } from "../package-metadata.mjs";
const i = e.forwardRef((m, g) => {
  K(B);
  const r = e.useRef(null);
  e.useImperativeHandle(
    g,
    () => ({
      element: r.current
    })
  );
  const {
    className: d,
    style: k,
    selected: n,
    disabled: a,
    item: x,
    render: y,
    dataItem: N,
    icon: p,
    svgIcon: b,
    text: u,
    id: C,
    onSelect: s,
    onKeyDown: l,
    index: o,
    tabIndex: E
  } = m, P = e.useMemo(
    () => D(
      "k-bottom-nav-item",
      {
        "k-selected": n,
        "k-disabled": a
      },
      d
    ),
    [n, a, d]
  ), R = e.useCallback(
    (c) => {
      s && o !== void 0 && !a && s(c, o);
    },
    [s, o, a]
  ), w = e.useCallback(
    (c) => {
      l && o !== void 0 && !a && l(c, o);
    },
    [l, o, a]
  ), v = y, I = x, f = /* @__PURE__ */ e.createElement(
    "span",
    {
      ref: r,
      className: P,
      style: k,
      role: "link",
      id: C,
      tabIndex: E,
      onClick: R,
      onKeyDown: w,
      "aria-current": n,
      "aria-disabled": a
    },
    I ? /* @__PURE__ */ e.createElement(I, { itemIndex: o, item: N }) : /* @__PURE__ */ e.createElement(e.Fragment, null, (p || b) && /* @__PURE__ */ e.createElement(h, { className: "k-bottom-nav-item-icon", name: p, icon: b, size: "xlarge" }), u && /* @__PURE__ */ e.createElement("span", { className: "k-bottom-nav-item-text", style: { userSelect: "none" } }, u))
  );
  return v !== void 0 ? v.call(void 0, f, m) : f;
}), M = {
  tabIndex: 0
};
i.propTypes = {
  className: t.string,
  style: t.object,
  id: t.string,
  disabled: t.bool,
  selected: t.bool,
  icon: t.string,
  svgIcon: T,
  text: t.string,
  tabIndex: t.number
};
i.defaultProps = M;
i.displayName = "KendoReactBottomNavigationItem";
export {
  i as BottomNavigationItem
};
