// @progress/kendo-react-layout/bottomnavigation/BottomNavigation.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import o from "prop-types";
import { validatePackage as P, focusFirstFocusableChild as T, useId as $, useDir as B, classNames as _, dispatchEvent as p, Keys as w } from "@progress/kendo-react-common";
import { packageMetadata as L } from "../package-metadata.mjs";
import { POSITION_MODE_CLASSES as A, ITEM_FLOW_CLASSES as H } from "./models/utils.mjs";
import { BottomNavigationItem as j } from "./BottomNavigationItem.mjs";
const b = t.forwardRef((a, I) => {
  P(L);
  const s = t.useRef(null), v = t.useCallback(
    () => {
      s.current && T(s.current);
    },
    []
  ), c = t.useCallback(
    () => ({
      element: s.current,
      focus: v
    }),
    [v]
  );
  t.useImperativeHandle(I, c);
  const {
    positionMode: k = i.positionMode,
    itemFlow: C = i.itemFlow,
    border: M = i.border,
    className: y,
    items: r,
    item: E,
    itemRender: N,
    disabled: m,
    style: O,
    id: h,
    onSelect: f,
    onKeyDown: g
  } = a, S = $(), F = B(s, a.dir), n = t.useMemo(
    () => a.fillMode === null || a.fill === null ? null : a.fill || a.fillMode || i.fillMode,
    [a.fillMode, a.fill]
  ), u = t.useMemo(
    () => a.themeColor || i.themeColor,
    [a.themeColor]
  ), D = t.useMemo(
    () => _(
      "k-bottom-nav",
      A[k],
      H[C],
      {
        [`k-bottom-nav-${n}`]: n,
        [`k-bottom-nav-${n}-${u}`]: !!(n && u),
        "k-bottom-nav-border": M,
        "k-disabled": m
      },
      y
    ),
    [k, u, n, C, M, m, y]
  ), d = t.useCallback(
    (e, l) => {
      r && !r[l].disabled && f && p(
        f,
        e,
        c(),
        {
          itemTarget: r[l],
          itemIndex: l
        }
      );
    },
    [r, f]
  ), R = t.useCallback(
    (e, l) => {
      d(e, l);
    },
    [d]
  ), K = t.useCallback(
    (e, l) => {
      switch (e.keyCode) {
        case w.enter:
        case w.space:
          d(e, l), e.preventDefault();
          break;
      }
      p(
        g,
        e,
        c(),
        void 0
      );
    },
    [d, g]
  );
  return /* @__PURE__ */ t.createElement(
    "nav",
    {
      ref: s,
      className: D,
      style: O,
      id: h || S,
      dir: F
    },
    r && r.map((e, l) => /* @__PURE__ */ t.createElement(
      j,
      {
        ...e,
        key: l,
        index: l,
        id: `${h || S}-${l}`,
        disabled: m || e.disabled,
        selected: e.selected,
        dataItem: e,
        item: E,
        render: N,
        onSelect: R,
        onKeyDown: K
      }
    ))
  );
}), i = {
  themeColor: "primary",
  fillMode: "flat",
  itemFlow: "vertical",
  positionMode: "fixed",
  border: !0
};
b.propTypes = {
  className: o.string,
  style: o.object,
  id: o.string,
  dir: o.string,
  themeColor: o.oneOf([
    "primary",
    "secondary",
    "tertiary",
    "info",
    "success",
    "warning",
    "error",
    "dark",
    "light",
    "inverse"
  ]),
  fill: o.oneOf(["solid", "flat"]),
  fillMode: o.oneOf(["solid", "flat"]),
  itemFlow: o.oneOf(["vertical", "horizontal"]),
  border: o.bool,
  disabled: o.bool,
  selected: o.number,
  onSelect: o.func
};
b.defaultProps = i;
b.displayName = "KendoReactBottomNavigation";
export {
  b as BottomNavigation
};
