// @progress/kendo-react-layout/splitter/SplitterPane.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as l from "react";
import { classNames as k } from "@progress/kendo-react-common";
class f extends l.Component {
  render() {
    const {
      id: t,
      size: e,
      collapsed: s,
      overlay: a,
      containsSplitter: o,
      collapsible: r,
      resizable: i,
      scrollable: n,
      keepMounted: c
    } = this.props, p = e && e.length > 0, d = {
      flexBasis: e
    }, m = k(
      "k-pane",
      {
        "k-hidden": s,
        hidden: s,
        "k-pane-flex": o,
        "k-pane-static": !i && !r || p,
        "k-scrollable": n
      }
    );
    return /* @__PURE__ */ l.createElement(
      "div",
      {
        id: t,
        role: "group",
        style: d,
        className: m
      },
      !s || c ? this.props.children : void 0,
      a ? /* @__PURE__ */ l.createElement("div", { className: "k-splitter-overlay k-overlay" }) : void 0
    );
  }
}
export {
  f as SplitterPane
};
