// @progress/kendo-react-layout/gridlayout/GridLayoutItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import t from "prop-types";
import { packageMetadata as I } from "../package-metadata.mjs";
import { validatePackage as R, useId as v } from "@progress/kendo-react-common";
const s = e.forwardRef((n, i) => {
  R(I);
  const a = e.useRef(null), l = e.useCallback(
    () => ({
      element: a.current
    }),
    []
  );
  e.useImperativeHandle(i, l);
  const { className: c, style: m, id: d, children: p, row: u, col: y, rowSpan: r, colSpan: o } = n, f = v(), g = {
    gridArea: `${u || "auto"} / ${y || "auto"} / ${r ? "span " + r : "auto"} / ${o ? "span " + o : "auto"}`,
    ...m
  };
  return /* @__PURE__ */ e.createElement(
    "div",
    {
      ref: a,
      className: c,
      style: g,
      id: d || f
    },
    p
  );
});
s.propTypes = {
  className: t.string,
  style: t.object,
  children: t.any,
  id: t.string
};
s.displayName = "KendoReactGridLayoutItem";
export {
  s as GridLayoutItem
};
