// @progress/kendo-react-layout/drawer/DrawerItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import t from "prop-types";
import { classNames as T, IconWrap as D, toIconName as P, getTabIndex as j, svgIconPropType as F } from "@progress/kendo-react-common";
import { DrawerContext as H } from "./context/DrawerContext.mjs";
const d = e.forwardRef((s, u) => {
  const {
    className: p,
    style: f,
    text: o,
    icon: n,
    svgIcon: c,
    separator: b,
    disabled: a,
    selected: I,
    onSelect: l,
    index: m,
    tabIndex: x,
    children: k,
    ...g
  } = s, { expanded: N, mini: w, item: C } = e.useContext(H), r = e.useRef(null), i = e.useCallback(
    () => {
      r.current && r.current.focus();
    },
    []
  );
  e.useImperativeHandle(u, () => ({
    element: r.current,
    focus: i,
    props: s
  }));
  const y = T(
    "k-drawer-item",
    {
      "k-selected": I,
      "k-disabled": a
    },
    p
  ), E = e.useCallback(
    (v) => {
      if (l && !a) {
        const R = {
          element: r.current,
          focus: i,
          props: s
        };
        l(R, m, v);
      }
    },
    [l, m, a]
  ), h = /* @__PURE__ */ e.createElement(e.Fragment, null, (n || c) && /* @__PURE__ */ e.createElement(D, { name: n && P(n), icon: c }), !N && w ? null : /* @__PURE__ */ e.createElement("span", { className: "k-item-text" }, o));
  return b ? /* @__PURE__ */ e.createElement("li", { className: "k-drawer-item k-drawer-separator", role: "separator" }) : /* @__PURE__ */ e.createElement(
    "li",
    {
      ref: r,
      className: y,
      style: f,
      role: "menuitem",
      "aria-label": o,
      "aria-disabled": a,
      onClick: E,
      tabIndex: j(x, a),
      ...g
    },
    C ? k : h
  );
});
d.propTypes = {
  className: t.string,
  style: t.object,
  text: t.string,
  icon: t.string,
  svgIcon: F,
  separator: t.bool,
  selected: t.bool,
  disabled: t.bool,
  targetItem: t.any
};
d.displayName = "KendoDrawerItem";
export {
  d as DrawerItem
};
