// @progress/kendo-react-layout/stepper/Step.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import a from "prop-types";
import { StepperContext as Z } from "./context/StepperContext.mjs";
import { focusFirstFocusableChild as ee, dispatchEvent as L, classNames as te, IconWrap as O, toIconName as ae } from "@progress/kendo-react-common";
import { useLocalization as se } from "@progress/kendo-react-intl";
import { checkOutlineIcon as ne, exclamationCircleIcon as oe } from "@progress/kendo-svg-icons";
import { DEFAULT_ANIMATION_DURATION as ce, NO_ANIMATION as le } from "./contants.mjs";
import { messages as ie, optionalText as re } from "./messages/index.mjs";
const g = e.forwardRef((R, V) => {
  const {
    // content
    children: z,
    className: E,
    current: p,
    disabled: n,
    focused: d,
    icon: i,
    svgIcon: h,
    index: s,
    isValid: t,
    label: r,
    optional: u,
    style: C,
    tabIndex: y,
    text: S,
    ...D
  } = R, {
    animationDuration: k,
    isVertical: b,
    item: H,
    linear: P,
    mode: F,
    numOfSteps: o,
    value: c,
    onChange: f,
    onFocus: I,
    successIcon: _,
    errorIcon: w,
    successSVGIcon: B,
    errorSVGIcon: G
  } = e.useContext(Z), m = e.useRef(null), A = e.useCallback(
    () => {
      m.current && ee(m.current);
    },
    []
  ), v = e.useCallback(
    () => ({
      element: m.current,
      focus: A
    }),
    [A]
  );
  e.useImperativeHandle(V, v);
  const N = !P || s === c - 1 || s === c || s === c + 1, M = F === "labels" || !!i && !!r, U = se(), W = ((l) => U.toLanguageString(l, ie[l]))(re), $ = typeof k == "number" ? k : k !== !1 ? ce : le, j = e.useCallback(
    (l) => {
      f && !n && L(
        f,
        l,
        v(),
        { value: s }
      );
    },
    [f, c, n]
  ), K = e.useCallback(
    (l) => {
      I && !n && L(
        I,
        l,
        v(),
        void 0
      );
    },
    [I, n]
  ), q = e.useMemo(
    () => te(
      "k-step",
      {
        "k-step-first": s === 0,
        "k-step-last": o && s === o - 1,
        "k-step-done": s < c,
        "k-step-current": p,
        "k-step-optional": u,
        "k-step-error": t !== void 0 && !t,
        "k-step-success": t
      },
      {
        "k-disabled": n,
        "k-focus": d
      },
      E
    ),
    [s, o, c, p, u, n, d, t, E]
  ), J = e.useMemo(
    () => ({
      maxWidth: b ? void 0 : `calc(100% / ${o})`,
      maxHeight: b ? `calc(100% / ${o})` : void 0,
      pointerEvents: N ? void 0 : "none",
      ...C
    }),
    [b, o, C, N]
  ), T = t ? _ : w, x = T ? /* @__PURE__ */ e.createElement(
    "span",
    {
      className: "k-step-indicator-icon " + T,
      "aria-hidden": "true"
    }
  ) : /* @__PURE__ */ e.createElement(
    O,
    {
      className: "k-step-indicator-icon",
      name: t ? "check-circle" : "exclamation-circle",
      icon: t ? B || ne : G || oe
    }
  ), Q = /* @__PURE__ */ e.createElement(e.Fragment, null, F !== "labels" ? /* @__PURE__ */ e.createElement(
    "span",
    {
      className: "k-step-indicator",
      "aria-hidden": !0,
      style: { transitionDuration: $ + "ms" }
    },
    i || h ? !M && t !== void 0 ? x : /* @__PURE__ */ e.createElement(O, { className: "k-step-indicator-icon", name: i && ae(i), icon: h }) : t !== void 0 ? x : /* @__PURE__ */ e.createElement("span", { className: "k-step-indicator-text" }, S || s + 1)
  ) : null), X = /* @__PURE__ */ e.createElement("span", { className: "k-step-label" }, r && /* @__PURE__ */ e.createElement("span", { className: "k-step-text" }, r), M && t !== void 0 && x, u && /* @__PURE__ */ e.createElement("span", { className: "k-step-label-optional" }, W)), Y = /* @__PURE__ */ e.createElement(e.Fragment, null, Q, X);
  return /* @__PURE__ */ e.createElement(
    "li",
    {
      ref: m,
      className: q,
      style: J,
      ...D
    },
    /* @__PURE__ */ e.createElement(
      "a",
      {
        className: "k-step-link",
        title: r || void 0,
        onClick: j,
        onFocus: K,
        tabIndex: y || (d ? 0 : -1),
        "aria-current": p,
        "aria-disabled": n || !N || void 0,
        "aria-invalid": t !== void 0 && !t || void 0
      },
      H ? z : Y
    )
  );
});
g.propTypes = {
  children: a.any,
  className: a.string,
  // content: PropTypes.any,
  current: a.bool,
  disabled: a.bool,
  icon: a.string,
  index: a.number,
  isValid: a.bool,
  label: a.string,
  optional: a.bool,
  style: a.object,
  tabIndex: a.number,
  text: a.string
};
const me = {
  tabIndex: 0
};
g.defaultProps = me;
g.displayName = "KendoStep";
export {
  g as Step
};
